fbw-a380x/src/systems/instruments/src/MFD/FMC/FmcAircraftInterface.tsx

```diff
@@ -803,15 +806,57 @@ export class FmcAircraftInterface {
       this.managedSpeedTargetIsMach = isMach;
     }
 
-    // Overspeed protection
-    const Vtap = Math.min(this.managedSpeedTarget ?? Vmo, SimVar.GetSimVarValue('L:A32NX_SPEEDS_VMAX', 'number'));
+    let Vtap = 0;
+    // VLS protection
+    if (this.managedSpeedTarget) {
+      const vls = this.speedVls.get();
+      if (this.managedSpeedTarget < vls) {
+        Vtap = vls;
+      }
+    }
 
+    if (!Vtap) {
+      // Overspeed protection
+      const vMax = SimVar.GetSimVarValue('L:A32NX_SPEEDS_VMAX', 'number');
+      Vtap = Math.min(this.managedSpeedTarget ?? Vmo - 5, vMax - 5);
+    }
     SimVar.SetSimVarValue('L:A32NX_SPEEDS_MANAGED_PFD', 'knots', vPfd);
     SimVar.SetSimVarValue('L:A32NX_SPEEDS_MANAGED_ATHR', 'knots', Vtap);
 
-    if (this.isAirspeedManaged()) {
+    const ismanaged = this.isAirspeedManaged();
+
+    if (ismanaged) {
       Coherent.call('AP_SPD_VAR_SET', 0, Vtap).catch(console.error);
     }
+
+    //short term managed speed
+    let shortTermManagedSpeed = 0;
+    const phase = this.fmgc.getFlightPhase();
+    if (phase != FmgcFlightPhase.Preflight) {
+      if (this.managedSpeedTarget) {
+        if (ismanaged) {
+          const verticalMode = SimVar.GetSimVarValue('L:A32NX_FMA_VERTICAL_MODE', 'number');
```

Comment:
Seems that these sorts of things should be published on the event bus in the FMS rather than having to read them directly in many places (general comment as there are a few new local var reads added).

Reply:
Thanks for the feedback. I've now moved some of the reads such as FMA vertical mode & Flight phase based on the event bus. The others I kept the same as they mostly seem to exists only be used this file and this logic should be on the PRIM anyways 😅